File: lib/geo.2d.circle.intersection.cc
enum circle_rel {
  SUB, SUP, // one contains the other
  TAN, // one intersection
  INT, // two intersections
  NOI // no intersection
};

circle_rel
circle_intersect(C a, C b) {
  Real d = Euclidean(a.o, b.o);
  if (d + a.r < b.r) return SUB; // a is in b
  if (a.r > d + b.r) return SUP; // b is in a
  if (d < a.r + b.r) return INT;
  if (d <= a.r + b.r + EPS) return TAN;
  return NOI;
}

File: procon/geo.2d.circle.intersection.rs
#[derive(Debug)]
enum CircleRelation {
    EQUAL,
    SUB, SUP,  // one contains the other
    INTERSECTION(usize)  // intersection with `n` points
}

fn circle_relation(a: Circle, b: Circle) -> CircleRelation {
    use CircleRelation::*;
    let d = (a.0 - b.0).norm();
    let eps = 1e-9;
    if d < eps && (a.1 - b.1).abs() < eps {
        EQUAL
    } else if d + a.1 < b.1 {
        SUB
    } else if a.1 > d + b.1 {
        SUP
    } else if d < a.1 + b.1 {
        INTERSECTION(2)
    } else if d <= a.1 + b.1 + eps {
        INTERSECTION(1)
    } else {
        INTERSECTION(0)
    }
}
